### PR TITLE
Implement basic soft logout handling

### PR DESCRIPTION
**Pull request by turt2live**
Fixes https://github.com/vector-im/riot-web/issues/10235
**Requires https://github.com/matrix-org/matrix-js-sdk/pull/975**

CSS and copy are left as an exercise for a later iteration.

Login page handling is left for https://github.com/vector-im/riot-web/issues/10236

This implementation reuses as much of the Lifecycle flow as it can without causing problems. Most importantly, it requires https://github.com/matrix-org/matrix-js-sdk/pull/975 to be able to detect a soft logout and react to it. When it comes time to starting/stopping the Lifecycle, additional parameters are provided so that the auxiliary services can (re)start themselves without the client starting to sync.

### CODE REVIEW

File: src/components/structures/MatrixChat.js [hunk 1]
```diff
@@ -1337,8 +1363,16 @@ export default React.createClass({
...
             }, true);
         });
-        cli.on('Session.logged_out', function(call) {
+        cli.on('Session.logged_out', function(errObj) {
             if (Lifecycle.isLoggingOut()) return;
+
+            if (errObj.httpStatus === 401 && errObj.data && errObj.data['soft_logout']) {
+                console.warn("Soft logout issued by server - avoiding data deletion");
+                Lifecycle.softLogout();
+                dis.dispatch({actions: 'soft_logout'});
```
Comment (bwindels) [inline src/components/structures/MatrixChat.js]: `action` with an "s"? Also, this action doesn't seem to be listened to anywhere else?

File: src/components/structures/MatrixChat.js [hunk 2]
```diff
@@ -1258,6 +1268,22 @@ export default React.createClass({
...
+     */
+    _onSoftLogout: function() {
+        this.notifyNewScreen('soft_logout');
+        this.setStateForNewView({
+            view: VIEWS.SOFT_LOGOUT,
+            ready: false,
+            collapseLhs: false,
+            collapsedRhs: false,
+            currentRoomId: null,
+            page_type: PageTypes.RoomDirectory,
```
Comment (bwindels) [inline src/components/structures/MatrixChat.js]: Not sure if `RoomDirectory` is the intended page type here, but if it is, I'd maybe put a comment here explaining why.

File: src/components/structures/MatrixChat.js [hunk 3]
```diff
@@ -432,6 +436,7 @@ export default React.createClass({
 
         switch (payload.action) {
             case 'logout':
+                console.log(payload);
```
Comment (bwindels) [inline src/components/structures/MatrixChat.js]: leftover logging

File: src/Lifecycle.js [hunk 4]
```diff
@@ -352,11 +352,14 @@ export function setLoggedIn(credentials) {
 async function _doSetLoggedIn(credentials, clearStorage) {
     credentials.guest = Boolean(credentials.guest);
 
+    const softLogout = isSoftLogout();
```
Comment (bwindels) [inline src/Lifecycle.js]: won't `clearStorage` still be `true` here when `Login` finishes your re-login through `setLoggedIn`?

Comment (turt2live) [reply src/Lifecycle.js]: This PR doesn't solve login. That's https://github.com/vector-im/riot-web/issues/10236

File: src/components/structures/MatrixChat.js (hunk 2 above)
Comment (turt2live) [reply src/components/structures/MatrixChat.js]: Copy/paste from above. I'm not sure what would be correct, or if it's even needed...

**Comment by turt2live** [comment]
it didn't format my cat when I removed it, so it's probably fine. I'll make a note to do more durable testing on it

**Pull request merged**